bpf/k_unix_sock.h

```diff
@@ -0,0 +1,298 @@
+        tcp_req_t *existing_tcp =
+            (tcp_req_t *)bpf_map_lookup_elem(&ongoing_tcp_req, &partial_p_conn);
+        if (existing_tcp) {
+            tcp_req_t *req = empty_tcp_req();
+            if (req) {
+                __builtin_memcpy(req, existing_tcp, sizeof(tcp_req_t));
+                bpf_map_delete_elem(&ongoing_tcp_req, &partial_p_conn);
+                bpf_map_update_elem(&ongoing_tcp_req, &p_conn, req, BPF_ANY);
+            }
+        } else {
+            http_info_t *existing_http =
+                (http_info_t *)bpf_map_lookup_elem(&ongoing_http, &partial_p_conn);
+
+            if (existing_http) {
+                http_info_t *req = empty_http_info();
+                if (req) {
+                    __builtin_memcpy(req, existing_http, sizeof(http_info_t));
+                    bpf_map_delete_elem(&ongoing_http, &partial_p_conn);
+                    bpf_map_update_elem(&ongoing_http, &p_conn, req, BPF_ANY);
+                }
+            }
+        }
+    }
+
+    // Make sure we don't have stale event from earlier socket connection if they are
+    // sent through the same socket. This mainly happens if the server overlays virtual
+    // threads in the runtime.
+    u64 sock_p = (u64)sk;
+    ensure_sent_event(id, &sock_p);
+
+    recv_args_t args = {
+        .sock_ptr = (u64)sk,
+    };
+
+    get_iovec_ctx((iovec_iter_ctx *)&args.iovec_ctx, msg);
+
+    bpf_map_update_elem(&active_recv_args, &id, &args, BPF_ANY);
+
+    return 0;
+}
+
+static __always_inline int return_unix_recvmsg(void *ctx, u64 id, int copied_len) {
+    recv_args_t *args = (recv_args_t *)bpf_map_lookup_elem(&active_recv_args, &id);
+
+    bpf_dbg_printk(
+        "=== return unix recvmsg id=%d args=%llx copied_len %d ===", id, args, copied_len);
+
+    if (!args || (copied_len <= 0)) {
+        return 0;
+    }
+
+    iovec_iter_ctx *iov_ctx = (iovec_iter_ctx *)&args->iovec_ctx;
+
+    if (!iov_ctx->iov && !iov_ctx->ubuf) {
+        bpf_dbg_printk("iovec_ptr found in kprobe is NULL, ignoring this tcp_recvmsg");
+        bpf_map_delete_elem(&active_recv_args, &id);
+
+        return 0;
+    }
+
+    struct sock *sock_ptr = (struct sock *)args->sock_ptr;
+
+    unsigned long inode_number;
+    unsigned long peer_inode_number = 0;
+    BPF_CORE_READ_INTO(&inode_number, sock_ptr, sk_socket, file, f_inode, i_ino);
+    struct unix_sock *usock = unix_sock_from_sk(sock_ptr);
+
+    if (usock) {
+        BPF_CORE_READ_INTO(&peer_inode_number, usock, peer, sk_socket, file, f_inode, i_ino);
+    }
+
+    bpf_dbg_printk("ino %d, peer ino %d", inode_number, peer_inode_number);
+
+    pid_connection_info_t p_conn = {.conn = {0}};
+
+    pid_connection_info_for_inode(id, &p_conn, inode_number, peer_inode_number);
+
+    bpf_map_delete_elem(&active_recv_args, &id);
+
+    u8 *buf = iovec_memory();
+    if (buf) {
+        copied_len = read_iovec_ctx(iov_ctx, buf, copied_len);
```

Comment:
nit: I'd not reuse the variable here, I found it a bit confusing - for instance, it begs the question: after a successful the call to read_iovec_ctx, should the new copied len match the older one? If not, why are they expected to be different?

Reply:
I can split it in another variable, but iovecs read with a loop and it's limited in how many iterations we do, so we may not read the full buffer.